src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -246,13 +235,25 @@ private void applyCleanedIndices(final ClusterChangedEvent event) {
     }
 
     private void applyDeletedIndices(final ClusterChangedEvent event) {
-        for (IndexService indexService : indicesService) {
-            final String index = indexService.index().name();
-            if (!event.state().metaData().hasIndex(index)) {
-                if (logger.isDebugEnabled()) {
-                    logger.debug("[{}] cleaning index, no longer part of the metadata", index);
-                }
+        final ClusterState previousState = event.previousState();
+        final String localNodeId = event.state().nodes().localNodeId();
+        assert localNodeId != null;
+
+        for (String index : event.indicesDeleted()) {
```

Comment:
indicesDeleted doesn't check for indexUUIDs. We have a separate method for it in this class `cleanMismatchedIndexUUIDs` - in this spirit of bringing all deletion code together - I think it's good to make indicesDelete aware of UUID switches (mark old as deleted), move the `applyDeletedIndices` to be executed where `cleanMismatchedIndexUUIDs` is called now and then we can remove `cleanMismatchedIndexUUIDs` make all go through here.